### ISSUE TITLE
Test/build all combinations of features

### ISSUE THREAD

**Issue by vorner** [rust-lang/cargo#4803]
I have already seen several bugs related to non-default selection of features (eg. it doesn't compile if `--no-default-features` is passed, something else fails to compile if a dependency gets additional feature through some other path in the dep graph).

I'm wondering if there's a way to help preventing these kinds of things. Some random ideas are:
* A flag to `cargo test` that would find all allowed combinations of features (or maybe even features of the dependencies) and iterate through that?
* Require this to pass on `cargo publish`?

But this is just an idea. I'm not sure this is the right place to start discussion, so if it's not, I'd be happy to be pointed to the right one.

**Comment by lukaslueg** [comment]
I've been caught by this as well and I think it would be useful if Cargo could manage to test all feature combinations automatically.

There are caveats, however: If a crate exports mutually exclusive features, the user must be able to control the invalid ones. Consider a crate that allows features `musl`, `libc`, `foo` and `bar`; it can be built with any or none of `foo` and `bar` and exactly one of `musl` or `libc` but not both. The user must be able to tell Cargo to build `(musl, )`, `(musl, foo)`, `(musl, bar)`, `(musl, foo, bar)`, `(libc, )`, `(libc, foo)`, `(libc, bar)` and `(libc, foo, bar)`.

To solve this, we could add  a flag `--feature-combinations` to `cargo test` that works in conjunction with `--features`, `--all-features` and `--no-default-features`.

* If `--feature-combinations` is supplied with `--features`, only the combinations of those features are built and tested
* If `--feature-combinations` goes alongside with `--all-features`, it's the same as with `--features` and an implicit list of all available features
* If `--feature-combinations` is supplied with `--no-default-features` an error is reported and the user called upon to select the desired basic list of features to be tested.

To drive home the example above, the user would have to do two invocations, one with `--features musl, foo, bar` and one with `--features libc, foo, bar`. We could allow multiple `--features`-arguments, I'm afraid what Cargo is about to do would quickly go beyond what the user can anticipate.

**Comment by lukaslueg** [comment]
I've been bitten by this once or twice and I think it's a useful feature to have in Cargo.

I'm wondering about how to provide the UI for this, though. Consider a crate that provides features `musl`, `libc`, `foo` and `bar`. None, one or both of `foo` and `bar` and exactly one of `musl` or `libc` can be specified; the crate is expected to and will fail to compile with none or both of `musl` and `libc` given.  The valid combinations therefor are `(musl, ), (musl, foo), (musl, foo, bar), (libc, ), (libc, foo) and (libc, foo, bar)`.
The point here is how to steer Cargo away from (rare) invalid combinations without having to explicitly state (plentiful, hard to get right) valid ones; we don't want a set-intersection-DSL on the commandline.

One way to solve this are two new flags `--feature-combinations` and `--all-feature-combinations` for `cargo test`. Given in conjunction with `--features` and `--default-features`, Cargo iterates over all unique combinations and builds+executes the test suite. The features given by `--features` are kept fixed, the ones given by `--feature-combinations` or discovered by `--all-feature-combinations` are mixed in.

To drive the example above home, the user would have to use two invocations, `cargo test --features libc --feature-combinations foo, bar; cargo test --features musl --feature-combinations foo, bar`.

A crate with no poisonous feature combinations can be tested by `cargo test --all-feature-combinations`, a subset of the default features (including the default features themselves) by `cargo test --default-features --all-feature-combinations`.

**Comment by luser** [comment]
The problem here is that you get combinatorial explosion in the number of features. There are 2<sup>n</sup> build configurations for `n` features. We've seen this countless times with the Firefox build system, where if we provide a `--disable-feature-x` option but don't have builds for it in CI it inevitably gets broken.

**Comment by vorner** [comment]
Yes, it's likely problematic with large code bases, because it takes long to compile just one combination, not saying much about 2²⁰.

But with a small crate with just 3 or 4 features, it should be possible to do as part of a single build (eg. not defining 16 different build jobs, but one that runs them linearly).

**Comment by jD91mZM2** [comment]
While we're waiting, a variation of [this script](https://gitlab.com/jD91mZM2/no-std-compat/blob/master/check.py) could be used for small and simple crates. Would really be awesome to have this as a built-in command though :)

**Comment by frewsxcv** [comment]
To those who may be interested, I just wrote a `test-all-features` Cargo subcommand that tests all feature combinations for a crate. You can check it out here: https://github.com/frewsxcv/cargo-test-all-features

<img src=https://i.imgur.com/OVBRtEC.png width=500>

**Comment by frewsxcv** [comment]
It also appears to be implemented in cargo-hack's `--feature-powerset` flag:

https://github.com/taiki-e/cargo-hack#usage

**Comment by Kinrany** [comment]
Checking all subsets is clearly impractical in the general case due to combinatorial explosion, so this likely needs to be handled in the compiler directly?

The compiler could accept one of three states for every feature declared implicitly with any `#[cfg(feature)]`: `on`, `off` or `optional`. It would then compile any item with an optional feature twice (or in the general case, once for every combination of optional features that apply to that item), with each compiled item annotated with the features.

The combinatorial explosion would still be there for individual items, but not for the whole crates.

Cargo and/or tools like `cargo-all-features` would then handle the configuration of the groups of features.

The necessary compiler feature should then perhaps be generalized to allowing for any compilation flag to be compiled as both `on` and `off` simultaneously. This would be very useful for `#[cfg(test)]` as well.
